ovs/dal/datalist.py

```diff
@@ -103,16 +116,132 @@ def guids(self):
             self._execute_query()
         return self._guids
 
+    def set_key(self, key=None):
+        """
+        Sets the caching key
+        Won't override the key when a key was giving on initializing
+        :param key: Key to explicitly use
+        :type key: str
+        :return: None
+        :rtype: NoneType
+        """
+        if key is not None:
+            self._key = '{0}_{1}'.format(DataList.NAMESPACE, key)
+            self._provided_key = True
+            # Unsure whether or not the same query would apply
+            self._volatile.delete(self._key)
+            return
+        elif self._provided_key is False:
+            identifier = copy.deepcopy(self._query)
+            identifier['object'] = self._object_type.__name__
+            identifier['guids'] = 'None' if self._provided_guids is None else ','.join(self._provided_guids)
+            self._key = '{0}_{1}'.format(DataList.NAMESPACE, hashlib.sha256(json.dumps(identifier)).hexdigest())
+
+    def _reset_list(self):
+        """
+        Resets everything about the DataList
+        :return: None
+        :rtype: NoneType
+        """
+        # Force query to rerun
+        self._executed = False
+        self._guids = None
+        self._data = {}
+        self._objects = {}
+        # Reset index information
+        self.from_index = 'none'
+        # Reset caching info
+        self._can_cache = True
+
+    def set_query(self, query):
+        """
+        Sets the query to apply to a different query
+        :param query: The query to perform. If this query is different from the previous one, the previously cached result will be replaced
+        with results of this query. If the query is identical and the result is cached, it will return the cached result
+        If None is supplied, a default query which does not filter anything will be set
+        :type query: dict or NoneType
+        Definitions:
+        * <query>: Should be a dictionary:
+                   {'type' : DataList.where_operator.XYZ,
+                    'items': <items>}
+        * <filter>: A tuple defining a single expression:
+                    (<field>, DataList.operator.XYZ, <value> [, <ignore_case>])
+                    The field is any property you would also find on the given object. In case of
+                    properties, you can dot as far as you like.
+        * <items>: A list of one or more <query> or <filter> items. This means the query structure is recursive and
+                   complex queries are possible
+        :return: None
+        :rtype: NoneType
+        """
+        if query is None:
```

Comment:
Don't know for sure whether `__init__` could be called here too. Since `reset_list` is called upon here
If possible, you don't need to validate the query, nor delete the volatile key

If not possible, you need to check whether the newly provided query is different from previously stored query
If not, no action is needed

Reply:
Not really required to check the query difference as the hash would be same and thus the caching value would be the same (covered in the unittest)
Deleting the volatile key is necessary when a key was specified as this key should always be used for this DataList instance. When not clearing the cache, changing the query or guids would return results from the previous _execute_query which is not what we'd want